annotation/tests/test_annotations.py

```diff
@@ -482,3 +593,665 @@ def test_construct_annotated_doc_db_error(
+            annotated_doc.tenant,
+            [{**page}, {**page}],
+        )
+
+        expected_result = ParticularRevisionSchema(
+            revision=annotated_doc.revision,
+            user=annotated_doc.user,
+            pipeline=annotated_doc.pipeline,
+            date=annotated_doc.date,
+            pages=[{**page}, {**page}],
+            validated=annotated_doc.validated,
+            failed_validation_pages=annotated_doc.failed_validation_pages,
+            categories=annotated_doc.categories,
+            similar_revisions=None,
+            links_json=annotated_doc.links_json,
+        )
+
+        assert result == expected_result
+
+
+def test_construct_particular_rev_response_error(annotated_doc: AnnotatedDoc):
+    with patch(
+        "annotation.annotations.main.load_annotated_pages_for_particular_rev",
+        side_effect=Exception,
+    ), patch(
+        "annotation.annotations.main.logger_.exception"
+    ) as mock_logger, pytest.raises(
+        Exception
+    ):
+        construct_particular_rev_response(annotated_doc)
+        mock_logger.assert_called_once_with("Can't load annotation")
+
+
+def test_check_null_fields(annotation_doc_for_save: DocForSaveSchema):
+    annotation_doc_for_save.pages = None
+    annotation_doc_for_save.validated = None
+    annotation_doc_for_save.failed_validation_pages = None
+
+    check_null_fields(annotation_doc_for_save)
+    assert annotation_doc_for_save.pages == []
+    assert annotation_doc_for_save.validated == set()
+    assert annotation_doc_for_save.failed_validation_pages == set()
+
+
+@pytest.mark.parametrize(("stop_revision"), ("not found", None))
+def test_accumulate_pages_info_stop_rev_none(
+    stop_revision: str, annotated_doc: AnnotatedDoc
+):
+    valid, fail, annot, not_proc, categ, required_rev = accumulate_pages_info(
+        task_pages=[1, 2, 3],
+        revisions=[annotated_doc],
+        stop_revision=stop_revision,
+        specific_pages=False,
+        with_page_hash=False,
+        unique_status=False,
+    )
+    assert valid == {1}
+    assert fail == set()
+    assert annot == set()
+    assert not_proc == {2, 3}
+    assert categ == annotated_doc.categories
+    assert required_rev is None
+
+
+@pytest.mark.parametrize(
+    ("stop_revision"), (LATEST, "20fe52cce6a632c6eb09fdc5b3e1594f926eea69")
+)
+def test_accumulate_pages_info_stop_rev_not_none(
+    stop_revision: str, annotated_doc: AnnotatedDoc
+):
+    valid, fail, annot, not_proc, categ, required_rev = accumulate_pages_info(
+        task_pages=[1, 2, 3],
+        revisions=[annotated_doc],
+        stop_revision=stop_revision,
+        specific_pages=False,
+        with_page_hash=False,
+        unique_status=False,
+    )
+    assert valid == {1}
+    assert fail == set()
+    assert annot == set()
+    assert not_proc == {2, 3}
+    assert categ == annotated_doc.categories
+    assert required_rev == annotated_doc
+
+
+def test_accumulate_pages_info_specific_pages_hash(
```

Comment:
You should consider mocking the getattr function, as it's used within this function, though it's not the main focus.